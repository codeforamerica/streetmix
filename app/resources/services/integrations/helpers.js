const { User } = require('../../../db/models')

/**
 finds the database record for the given user
 */
// might be better to do something besides return null
// we can actually probably get rid of this function if we want, but it might help
// keep some of the other code clean
const findUser = async function (userId) {
  const user = await User.findOne({ where: { auth0Id: userId } })
  if (user === null) {
    return null
  }
  return user.dataValues
}

// if the provider we're trying to find exists
// return its index value or -1 if not found
const addOrUpdateByProviderName = function (array, item) {
  const indexValue = array.findIndex(
    (_item) => _item.provider === item.provider
  )
  if (indexValue === -1) {
    array.push(item)
  } else {
    array[indexValue] = item
  }
}

/**
 given a user, check their identity data,
 depending on what data you find, add subscriber role

 TODO: make this less permissive(maybe add a status field or have logic depending on the provider), tested,
 and remove the role if we're confident we should (probably shouldn't worry about this for MVP)
 */
const syncAccountStatus = async function (userId) {
  const databaseUser = await User.findOne({ where: { auth0Id: userId } })
  if (databaseUser.identities.some((identity) => 'user_id' in identity)) {
    databaseUser.addRole('SUBSCRIBER_1')
  }
}

module.exports = { findUser, addOrUpdateByProviderName, syncAccountStatus }
