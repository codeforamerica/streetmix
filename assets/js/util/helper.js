function RandomGenerator() {
  this.randSeed = 0;
}

RandomGenerator.prototype.rand = function() {
  var t32 = 0x100000000;
  var constant = 134775813;
  var x = (constant * this.randSeed + 1);
  return (this.randSeed = x % t32) / t32;
}

RandomGenerator.prototype.seed = function(seed) {
  this.randSeed = seed;
}

function htmlEncode(value){
  //create a in-memory div, set it's inner text(which jQuery automatically encodes)
  //then grab the encoded contents back out.  The div never exists on the page.
  return $('<div/>').text(value).html();
}


function msg(messageId, data) {
  if (data) {
    return MESSAGES[messageId].supplant(data);
  } else {
    return MESSAGES[messageId];
  }
}

String.prototype.supplant = function(o) {
  return this.replace(/\[\[([^\[\]]*)\]\]/g,
    function (a, b) {
      var r = o[b];
      return typeof r === 'string' || typeof r === 'number' ? r : a;
    }
  );
};

function _createTimeout(fn, data, delay) {
  window.setTimeout(function() { fn.call(null, data); }, delay);
}

function _removeElFromDom(el) {
  if (el && el.parentNode) {
    el.parentNode.removeChild(el);
  }
}

function _getElAbsolutePos(el) {
  var pos = [0, 0];

  do {
    pos[0] += el.offsetLeft + (el.cssTransformLeft || 0);
    pos[1] += el.offsetTop + (el.cssTransformTop || 0);

    el = el.offsetParent;
  } while (el);

  return pos;
}

function _clone(obj) {
  if ($.isArray(obj)) {
    return $.extend(true, [], obj);
  } else {
    return $.extend(true, {}, obj);
  }
}
