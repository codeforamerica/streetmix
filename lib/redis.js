const redis = require('redis')
const config = require('config')
const logger = require('./logger.js')()

const initRedisClient = function () {
  let client, redisInfo

  if (config.redis.url) {
    redisInfo = new URL(config.redis.url)
    client = redis.createClient(redisInfo.port, redisInfo.hostname)
  } else {
    client = redis.createClient(config.redis.port, 'localhost')
  }

  const closeRedisConnection = function (error, exitCode) {
    if (error) {
      logger.error(error)
    }

    client.quit()
    client.on('end', function () {
      logger.info('Disconnected from Redis')
    })
  }

  // We do not want too many connections being made to Redis (especially for Streetmix production),
  // so before a process exits, close the connection to Redis.
  process.on('beforeExit', closeRedisConnection)
  process.on('SIGINT', closeRedisConnection)
  process.on('uncaughtException', closeRedisConnection)

  client.on('error', closeRedisConnection)

  client.on('connect', function () {
    logger.info('Connected to Redis')

    // Use the password in the URL if provided; otherwise use the one provided by config
    const redisAuth = (config.redis.url && redisInfo) ? redisInfo.password : config.redis.password

    if (redisAuth) {
      client.auth(redisAuth, function (error) {
        if (error) throw error
      })
    }
  })

  return client
}

module.exports = initRedisClient
