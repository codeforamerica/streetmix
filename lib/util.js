exports.parseLoginToken = function (req) {
  // authHeader is a string that looks like this:
  // `Streetmix realm="" loginToken="xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxx" userId="user"`
  const authHeader = req.header('Authorization')

  if (!authHeader) {
    return null
  }

  const auth = authHeader.split(/\s+/)

  if (auth.length < 4) {
    return null
  }

  const authScheme = auth.shift()

  // Convert the auth string to an object
  const authObject = auth.reduce((accumulator, item) => {
    const [ key, value ] = item.split('=')
    accumulator[key] = value.replace(/(^"|"$)/g, '')
    return accumulator
  }, {})

  if ((authScheme !== 'Streetmix') || (authObject.realm !== '')) {
    return null
  }

  return authObject
}

exports.ERRORS = {
  USER_NOT_FOUND: 'USER_NOT_FOUND',
  INTERNAL_ERROR: 'INTERNAL_ERROR',
  STREET_NOT_FOUND: 'STREET_NOT_FOUND',
  STREET_DELETED: 'STREET_DELETED',
  UNAUTHORISED_ACCESS: 'UNAUTHORISED_ACCESS',
  FORBIDDEN_REQUEST: 'FORBIDDEN_REQUEST',
  CANNOT_CREATE_STREET: 'CANNOT_CREATE_STREET',
  CANNOT_UPDATE_STREET: 'CANNOT_UPDATE_STREET',
  CANNOT_GET_STREET: 'CANNOT_GET_STREET',
  CANNOT_GET_USER: 'CANNOT_GET_USER'
}

exports.SAVE_THUMBNAIL_EVENTS = {
  INITIAL: 'INITIAL',
  SHARE: 'SHARE',
  TIMER: 'TIMER',
  BEFOREUNLOAD: 'BEFOREUNLOAD',
  PREVIOUS_STREET: 'PREVIOUS_STREET',
  TEST: 'TEST'
}

exports.requestIp = function (req) {
  if (req.headers['x-forwarded-for'] !== undefined) {
    return req.headers['x-forwarded-for'].split(', ')[0]
  } else {
    return req.connection.remoteAddress
  }
}
