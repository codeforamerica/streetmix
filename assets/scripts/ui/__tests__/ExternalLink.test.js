/* eslint-env jest */
import React from 'react'
import { renderWithReduxAndIntl } from '../../../../test/helpers/render'
import ExternalLink from '../ExternalLink'
import { ERRORS } from '../../../../lib/util'

const initialStateForOnline = {
  system: {
    noInternet: false
  }
}

const initialStateForOffline = {
  system: {
    noInternet: true
  }
}

describe('ExternalLink', () => {
  beforeAll(() => {
    global.window = Object.create(window)

    Object.defineProperty(window, 'location', {
      value: {
        hostname: 'streetmix.net'
      }
    })
  })

  it('renders an <a> element with string child', () => {
    const wrapper = renderWithReduxAndIntl(
      <ExternalLink href="https://example.com">foo</ExternalLink>,
      { initialState: initialStateForOnline }
    )
    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('renders an <a> element with element child', () => {
    const wrapper = renderWithReduxAndIntl(
      <ExternalLink href="https://example.com">
        <div>foo</div>
      </ExternalLink>,
      { initialState: initialStateForOnline }
    )
    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('renders string child without an <a> element in "no-internet" mode', () => {
    const wrapper = renderWithReduxAndIntl(
      <ExternalLink href="https://example.com">foo</ExternalLink>,
      { initialState: initialStateForOffline }
    )
    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('renders element child without an <a> element in "no-internet" mode', () => {
    const wrapper = renderWithReduxAndIntl(
      <ExternalLink href="https://example.com">
        <div>foo</div>
      </ExternalLink>,
      { initialState: initialStateForOffline }
    )
    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('should throw an error if href is not external', () => {
    expect(
      () => renderWithReduxAndIntl(
        <ExternalLink href="https://streetmix.net">
          <div>foo</div>
        </ExternalLink>,
        { initialState: initialStateForOnline }
      )
    ).toThrow(ERRORS.INVALID_EXTERNAL_LINK)
  })
})
