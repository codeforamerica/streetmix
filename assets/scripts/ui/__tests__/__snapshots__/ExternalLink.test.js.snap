// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExternalLink renders an <a> element with element child 1`] = `
<DocumentFragment>
  <a
    href="https://example.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    <div>
      foo
    </div>
  </a>
</DocumentFragment>
`;

exports[`ExternalLink renders an <a> element with string child 1`] = `
<DocumentFragment>
  <a
    href="https://example.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    foo
  </a>
</DocumentFragment>
`;

exports[`ExternalLink renders element child without an <a> element in "no-internet" mode 1`] = `
<DocumentFragment>
  <div>
    foo
  </div>
</DocumentFragment>
`;

exports[`ExternalLink renders string child without an <a> element in "no-internet" mode 1`] = `
<DocumentFragment>
  foo
</DocumentFragment>
`;
